src/mscorlib/src/System/Collections/Generic/DefaultObjectPool.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Collections.Generic
+{
+    internal class DefaultObjectPool<T> : ObjectPool<T> where T : class, new()
+    {
+        /// <summary>The default maximum number of oebjects that are available for rent.</summary>
+        private const int DefaultMaxPooled = 128;
+
+        private readonly T[] _objects;
+
+        private SpinLock _lock; // do not make this readonly; it's a mutable struct
+        private int _index;
+
+        /// <summary>
+        /// Creates the pool with DefaultMaxPooled objects.
+        /// </summary>
+        public DefaultObjectPool()
+            : this(DefaultMaxPooled)
+        {
+        }
+
+        /// <summary>
+        /// Creates the pool with maxPooled objects.
+        /// </summary>
+        public DefaultObjectPool(int maxPooled)
+        {
+            if (maxPooled <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxPooled));
+            }
+
+            _lock = new SpinLock(Debugger.IsAttached); // only enable thread tracking if debugger is attached; it adds non-trivial overheads to Enter/Exit
+            _objects = new T[maxPooled];
+        }
+
+        /// <summary>Takes an object from the pool. If the pool is empty, returns a new object.</summary>
+        public override T Rent()
+        {
+            T[] objects = _objects;
+            T obj = null;
+
+            // While holding the lock, grab whatever is at the next available index and
+            // update the index.  We do as little work as possible while holding the spin
+            // lock to minimize contention with other threads.  The try/finally is
+            // necessary to properly handle thread aborts on platforms which have them.
+            bool lockTaken = false;
+            try
+            {
+                _lock.Enter(ref lockTaken);
+
+                if (_index < objects.Length)
+                {
+                    obj = objects[_index];
+                    objects[_index++] = null;
+                }
+            }
+            finally
+            {
+                if (lockTaken) _lock.Exit(false);
+            }
+
+            return obj ?? new T();
```

Comment:
`new T()` will use `System.Activator` to create new instances. So i guess this will slow down things if the pool hast not yet been populated or if all items are in use and new ones are created. (Although I must admit that this is speculative, I have not measured this). Anyway, maybe provide a cached  `Func<T>` to the `Rent`method, so that it can create new objects without falling back to activator?

```csharp
public override T Rent(Func<T> activator)
{
  ....
return obj ?? activator();
}

// List<T> Code:
IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() 
{
    var enumerator = ObjectPool<PooledIEnumerator>.Shared.Rent(() => new PooledIEnumerator()); // delegate should be cached in a static field.
    return enumerator.Initalize(this);
}
```